# LCWA Powerpoint Dataset Exploration

This notebook uses python to explore the dataset of 1,000 ppt/x files
that were randomly identified and isolated from files archived by the
Library of Congress from US Federal government websites. 

In [41]:
import csv
import requests
import json
import collections
import pandas as pd

_NB: this file is delimited by tabs and in UTF-16 text encoding. 
You will need to specify these things, as demonstrated below, in order
for the file's data to be processable by python._ 

In [2]:
linecount = 0
with open('lcwa_gov_powerpoint_metadata.csv', 'r', newline='', encoding='utf-16') as data:
    lcwa_powerpoint_metadata = csv.reader(data, dialect='excel-tab', delimiter='\t')
    for line in lcwa_powerpoint_metadata:
        linecount += 1
        if linecount <= 3:
            print(linecount,' '.join(line))

1 urlkey timestamp original mimetype digest title company creation_date last_modified revision_number slide_count word_count file_size sha256 sha512
2 ada.ky.gov/adaemploymenttaxincentives.ppt 20100331231723 http://ada.ky.gov/ADAEMPLOYMENTTaxIncentives.ppt application/vnd.ms-powerpoint 4Y6Z6KRRHS3TIVQ2ZJVYVW7RSGXOBS6Y    Human Development Institute Promoting Independence, Productivity, and Integration for All People _________________________________    - 1998-12-17T14:31:23Z 2005-03-25T15:56:39Z 47 31 781 137216 9b6b74447622ac1646a538e63e452569e111cf775d1244373678d947405e33be 9567992755c0b1b38fdccb4440a4bb4b1758301b5cd1978b95bfe807babb70bf4c1e5ede85d9d26daee75196a3d62453ac9919bdbbf829508a26e72c35fc4df6
3 astrophysics.gsfc.nasa.gov/conferences/supernova1987a/staveley-smith.ppt 20090419154246 http://astrophysics.gsfc.nasa.gov/conferences/supernova1987a/Staveley-Smith.ppt application/vnd.ms-powerpoint 7EZSHVTKZE2FXP5SXI6DO2RJPIDLMXOP xNTD and the Pathway to SKA Lister Staveley-Smith - 200

In [3]:
print(linecount)

1001


In the cell above, we receive 1,001 as the line count since the variable indicates a count for all the lines, including the headers. So, this is what we would expect if the set includes 1,000 entries.

Now, let's read the dataset into a dictionary for easier processing.

In [4]:
lcwa_powerpoint_info = list()

with open('lcwa_gov_powerpoint_metadata.csv', 'r', newline='', encoding='utf-16') as data:
    lcwa_powerpoint_data = csv.DictReader(data, dialect='excel-tab', delimiter='\t')
    for line in lcwa_powerpoint_data:
        lcwa_powerpoint_info.append(line)

In [5]:
lcwa_powerpoint_info[:2]

[OrderedDict([('urlkey', 'ada.ky.gov/adaemploymenttaxincentives.ppt'),
              ('timestamp', '20100331231723'),
              ('original', 'http://ada.ky.gov/ADAEMPLOYMENTTaxIncentives.ppt'),
              ('mimetype', 'application/vnd.ms-powerpoint'),
              ('digest', '4Y6Z6KRRHS3TIVQ2ZJVYVW7RSGXOBS6Y'),
              ('title',
               '   Human Development Institute Promoting Independence, Productivity, and Integration for All People _________________________________   '),
              ('company', '-'),
              ('creation_date', '1998-12-17T14:31:23Z'),
              ('last_modified', '2005-03-25T15:56:39Z'),
              ('revision_number', '47'),
              ('slide_count', '31'),
              ('word_count', '781'),
              ('file_size', '137216'),
              ('sha256',
               '9b6b74447622ac1646a538e63e452569e111cf775d1244373678d947405e33be'),
              ('sha512',
               '9567992755c0b1b38fdccb4440a4bb4b1758301b5cd1978b9

In [30]:
pptx_headers = list()
for element in lcwa_powerpoint_info[0]:
    pptx_headers.append(element)

In [31]:
pptx_headers

['urlkey',
 'timestamp',
 'original',
 'mimetype',
 'digest',
 'title',
 'company',
 'creation_date',
 'last_modified',
 'revision_number',
 'slide_count',
 'word_count',
 'file_size',
 'sha256',
 'sha512',
 'status_code']

In [8]:
status_count = 0 

# see if sites return a live signal
for address in lcwa_powerpoint_info:
    try:
        url = address['original']
        r = requests.get(url)
        status_count += 1
        print(status_count, ':', r.status_code)
        address['status_code'] = r.status_code
    except:
        address['status_code'] = 'undetermined'
        status_count += 1
        print(status_count, ': undetermined')

200
200
200
200
200
404
200
200
404
200
404
200
404
404
404
404
200
200
200
200
200
200
404
404
404
404
200
200
200
200
200
200
200
404
200
404
404
200
200
404
200
200
200
200
200
200
404
200
404
404
404
200
200
200
404
404
404
200
200
200
404
200
404
404
200
200
200
200
200
200
200
406
200
404
200
200
404
404
200
200
200
200
404
404
200
200
404
200
200
200
200
200
200
404
200
200
200
200
404
404
404
404
404
404
404
404
404
200
404
404
404
404
200
200
404
404
404
200
404
200
404
404
404
404
200
200
404
200
200
200
200
200
200
404
200
200
404
404
200
200
200
200
404
404
404
200
200
200
404
200
404
404
403
200
200
200
404
404
404
404
404
200
200
200
404
200
404
404
200
404
404
200
200
200
200
200
404
200
200
200
200
200
200
404
404
404
401
404
404
200
404
404
200
404
404
404
200
503
200
200
404
403
200
404
404
404
404
200
200
404
403
404
200
200
200
500
200
200
200
404
404
404
404
200
200
404
404
200
200
404
404
404
200
200
404
200
200
404
200
200
200
200
200
200
200
200
200
200
404
404


In [9]:
lcwa_powerpoint_info[25]

OrderedDict([('urlkey', 'gsa.gov/graphics/admin/ae_acquisition_process_2.ppt'),
             ('timestamp', '20101128073912'),
             ('original',
              'http://www.gsa.gov/graphics/admin/AE_Acquisition_Process_2.ppt'),
             ('mimetype', 'application/vnd.ms-powerpoint'),
             ('digest', 'B4LNXAVPPNSQF4R3WBKCEQJEGWJF6JDV'),
             ('title', 'Slide 1'),
             ('company', 'GSA'),
             ('creation_date', '2007-11-23T17:13:53Z'),
             ('last_modified', '2009-05-28T20:27:46Z'),
             ('revision_number', '23'),
             ('slide_count', '29'),
             ('word_count', '1390'),
             ('file_size', '3301888'),
             ('sha256',
              'd1e33f6e63121f8281c8934c334163bdb95faa543881cf5b27a491c515ac0704'),
             ('sha512',
              '8051d0a78216cda488cacb37a846b4acaf427057823326ce993469a32f77034add45b4fe8f27e33b350ad13e02e072b96bd394388fd109aac3babbdc8bf57e44'),
             ('status_code', 404)])

In [11]:
status_count_check = 0

for item in lcwa_powerpoint_info:
    if item['status_code']:
        status_count_check += 1

print(status_count_check)

1000


In [32]:
#write the status codes into a data output
csvname = 'lcwa_gov_powerpoint_metadata_with_status-offsite-20190816.csv'
headers = list()
# uncomment below to add 'status_code' field to headers
#pptx_headers.append('status_code') #only run this once or check headers carefully

headers = pptx_headers #[:-4]
#headers.append('status_code')

print(headers)

['urlkey', 'timestamp', 'original', 'mimetype', 'digest', 'title', 'company', 'creation_date', 'last_modified', 'revision_number', 'slide_count', 'word_count', 'file_size', 'sha256', 'sha512', 'status_code']


In [33]:
with open(csvname, 'w', newline='', encoding='utf-8') as csvfile:
    writer = csv.DictWriter(csvfile, fieldnames=headers)
    writer.writeheader()    
    for item in lcwa_powerpoint_info:
        writer.writerow(item)
    print('wrote csv')
    

wrote csv


The below cell was intended as a check on non-responsive URLs, but that was corrected above. 

In [25]:
update_status_count = 0

# double check the nonresponsive sites
for address in lcwa_powerpoint_info:
    if address['status_code'] == 'undetermined':
        url = address['original']
        r = requests.get(url)
        address['status_code'] = r.status_code
        update_status_count += 1
        print(status_count,'Updating',url,'response:',r.status_code)
    else:
        continue
print('Updated status for',update_status_count,'items.')

SSLError: ("bad handshake: Error([('SSL routines', 'ssl3_get_server_certificate', 'certificate verify failed')],)",)

In [34]:
creators = dict()

for ppt in lcwa_powerpoint_info:
    company = ppt['company']
    if company not in creators:
        creators[company] = 1
    else:
        creators[company] += 1

print(creators)

{'-': 210, 'Lister Staveley-Smith': 1, 'State of Delaware': 1, 'SC Office of Human Resources': 1, 'Fermilab PPD': 1, 'CERN': 1, 'HRSA': 1, 'CDC': 8, 'NCHSTP': 1, 'UC Davis': 1, 'ICPSR': 1, '': 24, 'Southface Energy Institute': 1, 'Accenture': 2, 'State of Washington, Department of Personnel': 5, 'GSA': 4, 'Virginia Beach Public Schools': 1, 'NIH': 5, 'DOT': 1, 'Fire Safety Section, FAA': 2, 'US Treasury - FMS': 1, 'Low + Associates': 1, 'Center for Health Statistics': 1, 'USDA - NRCS': 1, 'DoED': 2, 'U.S. Department of Education': 2, 'Fermilab': 15, 'VITA': 1, 'DOER': 1, 'Bull Services': 1, 'St. John Evangelist': 1, 'Department of Communications': 1, 'Home': 1, 'LDCT': 1, 'USDA': 4, 'Tippecanoe County': 1, '<DIT>': 1, 'SLAC': 1, 'NASA/GSFC': 2, 'LAP-AUTH': 1, 'California Army National Guard': 13, 'Lockheed Martin Information Technology': 1, 'Department of Consumer Affair': 1, 'State of California': 3, 'GSFC/NASA': 1, 'Hooper Graphic Design/Perfect Mix': 1, 'Dell Computer Corporation': 

In [35]:
creators_sorted = collections.OrderedDict(creators)

items = list(creators_sorted)
items, len(items)

(['-',
  'Lister Staveley-Smith',
  'State of Delaware',
  'SC Office of Human Resources',
  'Fermilab PPD',
  'CERN',
  'HRSA',
  'CDC',
  'NCHSTP',
  'UC Davis',
  'ICPSR',
  '',
  'Southface Energy Institute',
  'Accenture',
  'State of Washington, Department of Personnel',
  'GSA',
  'Virginia Beach Public Schools',
  'NIH',
  'DOT',
  'Fire Safety Section, FAA',
  'US Treasury - FMS',
  'Low + Associates',
  'Center for Health Statistics',
  'USDA - NRCS',
  'DoED',
  'U.S. Department of Education',
  'Fermilab',
  'VITA',
  'DOER',
  'Bull Services',
  'St. John Evangelist',
  'Department of Communications',
  'Home',
  'LDCT',
  'USDA',
  'Tippecanoe County',
  '<DIT>',
  'SLAC',
  'NASA/GSFC',
  'LAP-AUTH',
  'California Army National Guard',
  'Lockheed Martin Information Technology',
  'Department of Consumer Affair',
  'State of California',
  'GSFC/NASA',
  'Hooper Graphic Design/Perfect Mix',
  'Dell Computer Corporation',
  'Department of Personnel',
  'sde',
  'Hewlett-P

In [36]:
sorted(creators_sorted.items())

[('', 24),
 ('\t閠]狴逄嬘뿿��', 1),
 ('\x19SAIC', 1),
 (' System 1, Inc.', 1),
 ('-', 210),
 ('20050404.MST', 1),
 ('586', 1),
 ('<DIT>', 1),
 ('ADNR/OPMP/ACMP', 1),
 ('AMS', 1),
 ('ANL', 1),
 ('APAC', 1),
 ('AS Division', 1),
 ('Accenture', 2),
 ('All Four Paws', 1),
 ('Argonne National Lab', 1),
 ('Argonne National Laboratory', 1),
 ('B&H', 1),
 ('BAE SYSTEMS', 1),
 ('BEA', 1),
 ('BNL', 4),
 ('BPA', 1),
 ('Beach-Self Determination', 1),
 ('Biology, University of Washington', 1),
 ('Black Graphics', 1),
 ('Booz Allen Hamilton', 1),
 ('Booz-Allen & Hamilton Inc.', 1),
 ('Boston University', 3),
 ('Brevard Community College', 2),
 ('Brookhaven National Laboratory', 3),
 ('Bull Services', 1),
 ('Bureau of Reclamation', 1),
 ('Byrd Polar Resaerch Center', 1),
 ('C-E', 3),
 ('CA Energy Commission', 1),
 ('CACI AB, Inc.', 1),
 ('CANG', 1),
 ('CCRS', 1),
 ('CCs', 1),
 ('CDC', 8),
 ('CDC/NCIRD', 1),
 ('CDC\\PHPPO\\OD', 1),
 ('CDIC', 1),
 ('CEC', 1),
 ('CERN', 1),
 ('CHIP', 1),
 ('CIMMS', 1),
 ('CP

In [37]:
sorted(creators_sorted.values(), reverse=True)

[210,
 25,
 24,
 15,
 15,
 13,
 13,
 11,
 9,
 8,
 8,
 7,
 6,
 6,
 5,
 5,
 5,
 5,
 5,
 5,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,

## Using pandas

This section was intended, but because it was expanded it is moved to a new notebook.

In [38]:
ppt_data = pd.read_csv('lcwa_gov_powerpoint_metadata.csv', encoding='utf-16', delimiter='\t', header=0, parse_dates=True)

In [39]:
ppt_df = pd.DataFrame.from_records(ppt_data)

In [40]:
ppt_df.head()

urlkey       timestamp  \
0          ada.ky.gov/adaemploymenttaxincentives.ppt  20100331231723   
1  astrophysics.gsfc.nasa.gov/conferences/superno...  20090419154246   
2  awm.delaware.gov/sitecollectiondocuments/awm+g...  20100220062758   
3  bcbintranet.sc.gov/ohr/hr-advisory/workforcepl...  20090416103442   
4  beamdocs.fnal.gov/ad/docdb/0010/001026/003/040...  20151115084934   

                                            original  \
0   http://ada.ky.gov/ADAEMPLOYMENTTaxIncentives.ppt   
1  http://astrophysics.gsfc.nasa.gov/conferences/...   
2  http://www.awm.delaware.gov/SiteCollectionDocu...   
3  http://www.bcbintranet.sc.gov/OHR/hr-advisory/...   
4  http://beamdocs.fnal.gov/AD/DocDB/0010/001026/...   

                        mimetype                            digest  \
0  application/vnd.ms-powerpoint  4Y6Z6KRRHS3TIVQ2ZJVYVW7RSGXOBS6Y   
1  application/vnd.ms-powerpoint  7EZSHVTKZE2FXP5SXI6DO2RJPIDLMXOP   
2  application/vnd.ms-powerpoint  ZHT4MJOX7D7OIRMCBPX4PMWZ4OQ7XSZ2   
3  application/vnd.ms-powerpoint  KDBPF6PJCBK2BEIYC2L5I33THZMINFML   
4  application/vnd.ms-powerpoint  OJOHGFEE4EI44UNSXLNV4Q5IKJ2L4H2Y   

                                               title  \
0     Human Development Institute Promoting Indep...   
1                        xNTD and the Pathway to SKA   
2                                   VSM Data Metrics   
3  Why Workforce Planning is Important during Har...   
4                            PowerPoint Presentation   

                        company         creation_date         last_modified  \
0                             -  1998-12-17T14:31:23Z  2005-03-25T15:56:39Z   
1         Lister Staveley-Smith                     -  2007-02-20T13:41:45Z   
2             State of Delaware  2006-10-16T16:19:52Z  2006-10-26T19:37:20Z   
3  SC Office of Human Resources  2009-01-07T14:28:04Z  2009-01-09T16:31:58Z   
4                  Fermilab PPD  2003-06-30T15:15:33Z  2004-02-22T22:33:55Z   

  revision_number slide_count word_count file_size  \
0              47          31        781    137216   
1              51          17        553   1580032   
2               4           1          2    136704   
3               7           9        238     53760   
4             122          45       1638   8227840   

                                              sha256  \
0  9b6b74447622ac1646a538e63e452569e111cf775d1244...   
1  34c77d3dc3cd3dc51d084584cafedf24189a42bf6bf6d3...   
2  1db9d640446ab52654f891458a30c6219e0125915f61c4...   
3  758efaac57fdf38d82bf7271f9cd5d1f083f842d884bff...   
4  3aeff996f66ee7f51e045ede1f5d21ab57639928b35804...   

                                              sha512  
0  9567992755c0b1b38fdccb4440a4bb4b1758301b5cd197...  
1  c36c59e694aad68c7b5fd4b32383e857520778131d9010...  
2  caa579649ea4a0a3126ab9842fe11e6e548695132d0da1...  
3  19449593a8c1defae33567966314bfe72d822b387eac6b...  
4  c6124015a81adf4d41e563b0a6460fd20275744f12b9b8...